File: apprefurbished/src/main/java/com/androidcollider/easyfin/transaction_categories/nested/TransactionCategoriesNestedModel.java
package com.androidcollider.easyfin.transaction_categories.nested;

import com.androidcollider.easyfin.common.models.TransactionCategory;
import com.androidcollider.easyfin.common.repository.Repository;

import java.util.List;

import io.reactivex.rxjava3.core.Flowable;

/**
 * @author Ihor Bilous
 */

class TransactionCategoriesNestedModel implements TransactionCategoriesNestedMVP.Model {

    private Repository repository;


    TransactionCategoriesNestedModel(Repository repository) {
        this.repository = repository;
    }


    @Override
    public Flowable<List<TransactionCategory>> getTransactionCategories(boolean isExpense) {
        return isExpense ?
                repository.getAllTransactionExpenseCategories() :
                repository.getAllTransactionIncomeCategories();
    }

    @Override
    public Flowable<TransactionCategory> updateTransactionCategory(TransactionCategory transactionCategory, boolean isExpense) {
        return isExpense ?
                repository.updateTransactionExpenseCategory(transactionCategory) :
                repository.updateTransactionIncomeCategory(transactionCategory);
    }

    @Override
    public Flowable<Boolean> deleteTransactionCategory(int id, boolean isExpense) {
        return isExpense ?
                repository.deleteTransactionExpenseCategory(id) :
                repository.deleteTransactionIncomeCategory(id);
    }
}

File: apprefurbished/src/main/java/com/androidcollider/easyfin/transaction_categories/nested/TransactionCategoriesNestedModel.kt
package com.androidcollider.easyfin.transaction_categories.nested

import com.androidcollider.easyfin.common.models.TransactionCategory
import com.androidcollider.easyfin.common.repository.Repository
import io.reactivex.rxjava3.core.Single

/**
 * @author Ihor Bilous
 */
internal class TransactionCategoriesNestedModel(private val repository: Repository) :
    TransactionCategoriesNestedMVP.Model {

    override fun getTransactionCategories(isExpense: Boolean): Single<List<TransactionCategory>> {
        return if (isExpense)
            repository.allTransactionExpenseCategories!!
        else repository.allTransactionIncomeCategories!!
    }

    override fun updateTransactionCategory(
        transactionCategory: TransactionCategory,
        isExpense: Boolean
    ): Single<TransactionCategory> {
        return if (isExpense)
            repository.updateTransactionExpenseCategory(transactionCategory)
        else repository.updateTransactionIncomeCategory(transactionCategory)
    }

    override fun deleteTransactionCategory(id: Int, isExpense: Boolean): Single<Boolean> {
        return if (isExpense)
            repository.deleteTransactionExpenseCategory(id)
        else repository.deleteTransactionIncomeCategory(id)
    }
}